src/executor/sort_merge_join.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+#![allow(dead_code)]
+use std::cmp::Ordering;
+use std::vec::Vec;
+
+use super::*;
+use crate::array::{DataChunk, DataChunkBuilder};
+use crate::types::{DataType, Row};
+/// The executor for sort merge join
+pub struct SortMergeJoinExecutor {
+    pub left_child: BoxedExecutor,
+    pub right_child: BoxedExecutor,
+    pub left_column_index: usize,
+    pub right_column_index: usize,
+    pub left_types: Vec<DataType>,
+    pub right_types: Vec<DataType>,
+}
+impl SortMergeJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut left_rows = row_stream(self.left_child);
+        let mut right_rows = row_stream(self.right_child);
+        let mut left_row;
+        let mut right_row;
+        let first_left_row = left_rows.next().await;
+        let first_right_row = right_rows.next().await;
+        if first_left_row.is_none() || first_right_row.is_none() {
+            yield Err(ExecutorError::NotNullable)?
+        } else {
+            left_row = first_left_row.unwrap()?;
+            right_row = first_right_row.unwrap()?;
+            // build
+            let data_types = self.left_types.iter().chain(self.right_types.iter());
+            let mut builder = DataChunkBuilder::new(data_types, PROCESSING_WINDOW_SIZE);
+
+            loop {
+                match compare_row(
+                    &left_row,
+                    &right_row,
+                    self.left_column_index,
+                    self.right_column_index,
+                ) {
+                    Ordering::Equal => {
```

Comment:
Still some little problem with `Equal`. Assuming you are doing an inner join. If we have:

left: [1, 1]
right: [1, 1]

The join result should be four lines of [1, 1]. However, the current implementation doesn't seem correct?

Please also add this test case, thanks!

Reply:
> The join result should be four lines of [1, 1]. However, the current implementation doesn't seem correct?

I found that the bug you mentioned was caused by the unit tests not being constructed correctly.
So I decided to make DataChunk contain two columns and added some test cases, and now it's okay.